core/trino-main/src/main/java/io/trino/operator/PagesSpatialIndexSupplier.java

```diff
@@ -60,7 +61,7 @@
     private final LongArrayList addresses;
     private final List<Type> types;
     private final List<Integer> outputChannels;
-    private final List<List<Block>> channels;
+    private final List<ObjectArrayList<Block>> channels;
```

Comment:
can we keep it as `List<Block>`. Generally we don't expose implementation of collection in fields and params

Reply:
It has to be specific to allow proper memory accounting. Since the collection is expanding it is important to account for the entire array and not only for the part containing blocks. In most cases it's probably not a big deal, however when the number of pages is high the overhead of dynamic arrays could be significant.

Reply:
Can we have some util method `listRetinedSize(List<T> list, Function<T,Long> elementSize)` which does `instanceof` on `list` and applies proper logic?

Reply:
I thought about that. The problem is that if the interface allows any type of list you need to enumerate any possible ones to ensure proper memory accounting. And if a previously unknown list passed you either have to make an assumption (what would result in an incorrect memory estimates) or to throw an exception. The first is what I'm actually trying to avoid in this PR, and the second is IMO worse than simply defining a specific interface.